Allow using /etc/hosts for RabbitMQ name resolution

dshcherb opened the pull request:
Allow using /etc/hosts for rmq name resolution
Add support for network spaces

dshcherb commented:
Pending functional testing.

dshcherb commented:
Pull-request updated, HEAD is now 60b061155235e113884922e7437c32d16b754c1a

dshcherb commented:
Pull-request updated, HEAD is now 9a29c00c63a4863fee7f3e99c1afe4a12884a158

Andrey-mp commented:
it would be good to split these commits in two different PR (and maybe fix typo in separate also).
rabbitmq issue is a hack that can be reverted later and it can be reviewed faster.
supporting of network spaces is more complicated and I need more time for this.

dshcherb commented:
Pull-request updated, HEAD is now 580dd537baebd04a2bee1eb29d91ce3711241717

Andrey-mp commented on contrail-controller/hooks/contrail_controller_hooks.py:
```diff
@@ -155,6 +172,9 @@ def config_changed():
         if is_leader():
             _address_changed(local_unit(), ip)
 
+        if config.get('local-rabbitmq-hostname-resolution'):
```
this will lead to updating /etc/hosts only with local IP/name. another logic is needed to be able to change this option on-the-fly.

Andrey-mp commented on contrail-controller/hooks/contrail_controller_utils.py:
```diff
@@ -179,3 +185,66 @@ def update_charm_status():
+
+    newlines = []
+    for ip, hostname in hosts_map.items():
+        if not ip or not hostname:
```
is it realistic that ip or hostname can be empty?

dshcherb replied:
Probably not, unless a caller does it by accident. We could raise an exception here or just skip as it is done here.

The code of this function was literally copied from here: https://github.com/openstack/charm-rabbitmq-server/blob/stable/18.11/hooks/rabbit_utils.py#L663-L700

Andrey-mp commented on contrail-controller/hooks/contrail_controller_utils.py:
```diff
@@ -179,3 +185,66 @@ def update_charm_status():
     common_utils.update_services_status(SERVICES)
+
+
+def update_hosts_file(hosts_map):
```
i see that all client pass just one item to this function - maybe it's better to change interface from host_map to (ip, hostname) ?

dshcherb replied:
Makes sense, will do.

Andrey-mp commented on contrail-controller/hooks/contrail_controller_utils.py:
```diff
@@ -179,3 +185,66 @@ def update_charm_status():
+            for line in lines:
+                hosts.write(line)
+
+    os.rename(tmpfile.name, HOSTS_FILE)
```
why did you make writing with renaming?
if this operation should be atomic - then it should be done in the same folder. rename is not atomic across different disks that can be in case /tmp and /etc.

dshcherb replied:
Just reused the code from rabbitmq-charm.

The original commit isn't very descriptive but I would assume that the idea is to make the replacement atomic to avoid the situation where application runtimes (glibc etc.) would be reading the hosts file while it is being modified by the charm process. 

https://github.com/openstack/charm-rabbitmq-server/commit/8272f214ab2190c17cf8c84f7cd9f227dcabdf47#diff-21870f9ed3cd89ae3ca8d1f237afc315R384

I can pass the dir argument to NamedTemporaryFile to cover the case you described.

Andrey-mp commented on contrail-controller/hooks/contrail_controller_utils.py:
```diff
@@ -179,3 +185,66 @@ def update_charm_status():
+
+    @returns hostname
+    """
+    return '{}-contrail-rmq'.format(socket.gethostname())
```
socket.gethostname() may return fqdn - should it be caught here?

Andrey-mp commented on contrail-controller/hooks/contrail_controller_utils.py:
```diff
@@ -179,3 +185,66 @@ def update_charm_status():
+                log("Removing line '%s' from hosts file" % (line))
+
+        lines = keepers
+        newlines.append("%s %s\n" % (ip, hostname))
```
is it sufficient to store only one form in /etc/hosts ? should it add short and fqdn forms?

Andrey-mp commented on contrail-controller/hooks/contrail_controller_utils.py:
```diff
@@ -179,3 +185,66 @@ def update_charm_status():
+        keepers = []
+        for line in lines:
+            _line = line.split()
+            if len(line) < 2 or not (_line[0] == ip or hostname in _line[1:]):
```
with this condition real name with this IP will be removed that is not good.

Andrey-mp commented:
1) Fixing /etc/hosts is not so good solution. Anyone else may re-write it or current code can break changes of someone else.
2) This solution may lead to inability to resolve some names that were stored in /etc/hosts and were removed due to current implementation.

dshcherb commented:
Pull-request updated, HEAD is now bef9c96844914aecc3aed89134941d93d9a850f1

dshcherb commented:
Thanks for the review.

I updated the PR (pending functional testing).

Previously used hostnames are now pushed to aliases (http://man7.org/linux/man-pages/man5/hosts.5.html) and are restored if the charm option to manage /etc/hosts is disabled (admittedly the previously used hostname will not be restored correctly if there are multiple aliases already as the previous hostname is appended to the list of aliases - I could fix that if this is important but it is an edge case). A persistent store (unitdata kv) is used to remember the managed `ip: hostname` pairs - this is used to disable the feature after the relevant config option is set to false.

dshcherb commented:
Pull-request updated, HEAD is now a578f6d977306fe51fc88858fe8775cbdb5b2850

dshcherb commented:
Pull-request updated, HEAD is now dc092ea3ed785d338cd4b4063b3f2f4966d493c9

dshcherb commented:
@Andrey-mp Pinged you there.

Also fixed 2 additional issues (see the commit message) and added more logging to track down the problem where one unit had only 2/3 hostnames although all events (joined and changed) for all (2) peer units ran for that unit.

I am doing another redeployment to see if I can reproduce that and see what is wrong (with more Juju and charm logging).

dshcherb commented:
Pull-request updated, HEAD is now e5ab1cb5286d64728630e9fddb000db4f8090e7b

dshcherb commented:
With https://github.com/Juniper/contrail-charms/commit/e5ab1cb5286d64728630e9fddb000db4f8090e7b I got to a good state in a functional test where all units had their /etc/hosts filled with all cluster ip:hostname pairs without any duplicates.

Hostnames appear on all machines after all controller-cluster-changed events fire.

Andrey-mp commented on contrail-controller/hooks/contrail_controller_hooks.py:
```diff
@@ -143,6 +175,16 @@ def config_changed():
         settings = {"private-address": ip}
+
+        if config.get('local-rabbitmq-hostname-resolution'):
+            settings.update({
```
this 'settings' is for other apps relations. settings for cluster relation is in line 194

dshcherb replied:
Thanks for catching - you are right.

Andrey-mp commented:
i don't have any requests in the skype...

about the change - right now it become rather big and not so simple. But from the other hand it's isolated. Therefore it can be treated as blackbox and if it works for you then I can merge it.

But in general I suppose that logic can be simplified with:
- store hostname in relation always. without checking is it enable in config or not. flag in config can be used just for switch to fill /etc/hosts or not. storing can be done in the same manner as unit-address. at the any point in time this list can be obtained from relation without storing in kv.
- because hostname is rather unique and predefined - then working with /etc/hosts can be simplified. in update_charm_status func - after reading we can remove all lines where this hostname is present (either as hostname or in aliases). and then code can read current state from relation and form new content for /etc/hosts. if new content is not equal to previous then new file can be renamed.

If this work is urgent and you confirm that it works for you - please let me know and i'll merge it. Later the change can be re-worked.

dshcherb commented:
Pull-request updated, HEAD is now d5b280282a71e464a1fe482eeae9c9a23a1ce17d

dshcherb commented:
Yes, from the perspective of the project I am working on it is quite critical as:

- OAM network interface is used for Juju communication;
- One of the secondary interfaces of a node is used for control plane & MX communication;
- Doing /etc/hosts management in a separate charm or a script introduces race conditions.

I managed to get to a deployment where contrail-status was good yesterday:

https://gist.github.com/dshcherb/7828b0704cc41c7f88f59ff6c4c689d1

So I think having it under a feature flag and using it to move forward is a workable solution until the docker image is fixed to use FQDNs.
